### PR TITLE
bpo-25324: copy tok_name before changing it

### PR DESCRIPTION
**Pull request by albertjan**
Saw this open bug report, and since I was looking at tokenize.py anyway. I figured I address it. 

This may catch people off guard though because they may be relying on `tok_name` containing `ENCODING`, `COMMENT` and `NL`. 🤷‍♂️

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Lib/test/test_tokenize.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from test import support
 from tokenize import (tokenize, _tokenize, untokenize, NUMBER, NAME, OP,
                      STRING, ENDMARKER, ENCODING, tok_name, detect_encoding,
-                     open as tokenize_open, Untokenizer)
+                     open as tokenize_open, Untokenizer, tok_name as tokenize_tok_name)
```
Comment (serhiy-storchaka) [inline Lib/test/test_tokenize.py]: Too long line.

File: Lib/test/test_tokenize.py [hunk 2]
```diff
@@ -1341,13 +1342,13 @@ def assertExactTypeEqual(self, opstr, *optypes):
         tokens = list(tokenize(BytesIO(opstr.encode('utf-8')).readline))
         num_optypes = len(optypes)
         self.assertEqual(len(tokens), 2 + num_optypes)
-        self.assertEqual(token.tok_name[tokens[0].exact_type],
-                         token.tok_name[ENCODING])
+        self.assertEqual(tokenize_tok_name[tokens[0].exact_type],
+                         tokenize_tok_name[ENCODING])
```
Comment (haypo) [inline Lib/test/test_tokenize.py]: I would prefer to import the tokenize module and write tokenize.tok_name.

I know that this file uses "from tokenize import ..." but this pattern is unusual in the Python test suite.

**Review by vstinner — CHANGES_REQUESTED** [review]
I'm not sure that it's the right way to fix the issue: see http://bugs.python.org/issue25324 discussion. I also have comments on the change itself, but I will want until we agree on the way to fix the issue before reviewing the change.

**Comment by brettcannon** [comment]
Due to a new release of Sphinx, we had to fix the documentation to build on Travis again. Please do a merge to get these changes to help get Travis passing on your PR.

File: Lib/token.py [hunk 3]
```diff
@@ -66,8 +66,11 @@
 OP = 53
 AWAIT = 54
 ASYNC = 55
-ERRORTOKEN = 56
-N_TOKENS = 57
+COMMENT = 56
```
Comment (vstinner) [inline Lib/token.py]: Please don't change ERRORTOKEN value.

File: Include/token.h [hunk 4]
```diff
@@ -66,8 +66,12 @@ extern "C" {
 #define OP		53
 #define AWAIT		54
 #define ASYNC		55
-#define ERRORTOKEN	56
-#define N_TOKENS	57
+/* These aren't used by the c tokenizer but are needed for tokenize.py */
```
Comment (vstinner) [inline Include/token.h]: Niypick: replace c with C.

File: Lib/token.py (hunk 3 above)
Comment (vstinner) [reply Lib/token.py]: Please copy token.h comment here.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by albertjan** [comment]
Updated the PR. comments are now copied from token.h to token.py automatically. And I moved ERRORTOKEN back to where it was.

File: Lib/test/test_tokenize.py [hunk 5]
```diff
@@ -1417,7 +1417,6 @@ def test_pathological_trailing_whitespace(self):
         # See http://bugs.python.org/issue16152
         self.assertExactTypeEqual('@          ', token.AT)
 
-
```
Comment (vstinner) [inline Lib/test/test_tokenize.py]: This change is not PEP 8 compliant :-)

File: Lib/token.py [hunk 6]
```diff
@@ -104,13 +110,23 @@ def _main():
     prog = re.compile(
         "#define[ \t][ \t]*([A-Z0-9][A-Z0-9_]*)[ \t][ \t]*([0-9][0-9]*)",
         re.IGNORECASE)
+    comment = re.compile(
+        "^\s*/\*\s*(.+)\s*\*/\s*$",
```
Comment (vstinner) [inline Lib/token.py]: This string emits:

```
<stdin>:1: DeprecationWarning: invalid escape sequence \s
```

I suggest to always use raw strings for regular expressions.

File: Lib/token.py [hunk 7]
```diff
@@ -104,13 +110,23 @@ def _main():
...
     tokens = {}
+    prev_val = None
     for line in lines:
         match = prog.match(line)
         if match:
             name, val = match.group(1, 2)
-            val = int(val)
-            tokens[val] = name          # reverse so we can sort them...
+            prev_val = int(val)
+            tokens[prev_val] = {'token': name}          # reverse so we can sort them...
```
Comment (vstinner) [inline Lib/token.py]: I would prefer to use val here, but set prev_val after the tokens assignement. Here we use the current value, not the previous value.

File: Lib/token.py [hunk 8]
```diff
@@ -104,13 +110,23 @@ def _main():
...
+        else:
+            comment_match = comment.match(line)
+            if comment_match and prev_val:
```
Comment (vstinner) [inline Lib/token.py]: "prev_val is not None" to support prev_val == 0 (ENDMARKER = 0).

File: Lib/token.py [hunk 9]
```diff
@@ -104,13 +110,23 @@ def _main():
...
+                val = comment_match.group(1)
```
Comment (vstinner) [inline Lib/token.py]: nitpick: i suggest to rename the variable "comment".

File: Lib/token.py [hunk 10]
```diff
@@ -128,7 +144,9 @@ def _main():
         sys.exit(3)
     lines = []
     for val in keys:
```
Comment (vstinner) [inline Lib/token.py]: nitpick: I suggest to rename "val" to "key" to be more consistent.

**Review by vstinner — COMMENTED** [review]
LGTM, but I added a new (last I hope) serie of nitpicking comments :-p

**Comment by albertjan** [comment]
This should address your comments. Thanks for taking the time to review my PR.

**Review by vstinner — APPROVED** [review]
LGTM. I will merge the change once tests pass.

**Comment by serhiy-storchaka** [comment]
Please wait with merging. I'm finishing my patch for generating `token.h` from `token.py`.

**Comment by vstinner** [comment]
> Please wait with merging. I'm finishing my patch for generating token.h from token.py.

Wait? Do you expect conflicts?

**Comment by serhiy-storchaka** [comment]
Yes, conflicts, and maybe this will lead to redesigning both patches.

File: Misc/NEWS [hunk 11]
```diff
@@ -10,6 +10,10 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- bpo-25324: Tokens needed for parsing in python moved to C. ``COMMENT``,
+  ``NL`` AND ``ENCODING``. This way the tokens and tok_names in token.py
```
Comment (vstinner) [inline Misc/NEWS]: and: lower case

in token.py: in the token module

import tokenize.py: import the tokenize module.

**Comment by albertjan** [comment]
Should I do a merge or a rebase to resolve the conflicts?

**Comment by vstinner** [comment]
> Should I do a merge or a rebase to resolve the conflicts?

As you want.

File: Misc/NEWS [hunk 12]
```diff
@@ -10,6 +10,10 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- bpo-25324: Tokens needed for parsing in python moved to C. ``COMMENT``,
```
Comment (serhiy-storchaka) [inline Misc/NEWS]: Python: title case.

File: Doc/library/token.rst [hunk 13]
```diff
@@ -101,10 +101,18 @@ The token constants are:
...
           N_TOKENS
           NT_OFFSET
 
    .. versionchanged:: 3.5
       Added :data:`AWAIT` and :data:`ASYNC` tokens. Starting with
       Python 3.7, "async" and "await" will be tokenized as :data:`NAME`
       tokens, and :data:`AWAIT` and :data:`ASYNC` will be removed.
+
+   .. versionchanged:: 3.7
+      Added :data:`COMMENT`, :data:`NL` and :data:`ENCODING`. To bring
```
Comment (serhiy-storchaka) [inline Doc/library/token.rst]: Isn't a period is redundant here?

File: Doc/library/token.rst [hunk 14]
```diff
@@ -101,10 +101,18 @@ The token constants are:
...
+      the tokens in the C code in line with the tokens needed in
+      tokenize.py. These tokens aren't used by the C tokenizer.
```
Comment (serhiy-storchaka) [inline Doc/library/token.rst]: tokenize.py -> the :mod:\`tokenize\` module

**Review by serhiy-storchaka — DISMISSED** [review]

**Comment by vstinner** [comment]
LGTM. Thanks @albertjan!

**Comment by ammaraskar** [comment]
Hey @albertjan, did you manually edit token.py for this change? As part of my pr here for testing the round tripping of https://github.com/python/cpython/pull/681, it looks like your commit eats up the last token if its followed by a comment.

i.e currently the file generated by token.py looks something like this
```python
ATEQUAL = 50
RARROW = 51
# Don't forget to update the table _PyParser_TokenNames in tokenizer.c!
OP = 53
AWAIT = 54
ASYNC = 55
# These aren't used by the C tokenizer but are needed for tokenize.py
COMMENT = 57
NL = 58
ENCODING = 59
```

whereas the original token.py looks like
```python
ATEQUAL = 50
RARROW = 51
ELLIPSIS = 52
# Don't forget to update the table _PyParser_TokenNames in tokenizer.c!
OP = 53
AWAIT = 54
ASYNC = 55
ERRORTOKEN = 56
# These aren't used by the C tokenizer but are needed for tokenize.py
COMMENT = 57
NL = 58
ENCODING = 59
```

Notice how `ERRORTOKEN` and `ELLIPSIS` haven't been added in the auto generated file. I'll fix it as part of my PR but just wondering if you manually edited the file.

**Comment by ammaraskar** [comment]
Nevermind @albertjan, looks like the issue I described here was caused by a merge issue. Thanks anyway :)

**Pull request merged**